controllers/clustertopology_controller_sync.go

```diff
@@ -99,3 +108,193 @@ func checkReferencedObjectsAreStrictlyCompatible(current, desired client.Object)
+	for _, mdID := range toCreate {
+		md := desiredMachineDeployments[mdID]
+		if err := r.createMachineDeployment(ctx, md); err != nil {
+			return err
+		}
+	}
+
+	// Delete MachineDeployments.
+	for _, mdID := range toDelete {
+		md := currentMachineDeployments[mdID]
+		if err := r.deleteMachineDeployment(ctx, md); err != nil {
+			return err
+		}
+	}
+
+	// Update MachineDeployments.
+	for _, mdID := range toUpdate {
+		currentMD := currentMachineDeployments[mdID]
+		desiredMD := desiredMachineDeployments[mdID]
+		if err := r.updateMachineDeployment(ctx, current.cluster.Name, currentMD, desiredMD); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *ClusterTopologyReconciler) createMachineDeployment(ctx context.Context, md machineDeploymentTopologyState) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	log.Info("creating", md.infrastructureMachineTemplate.GroupVersionKind().String(), md.infrastructureMachineTemplate.GetName())
+	if err := r.Client.Create(ctx, md.infrastructureMachineTemplate); err != nil {
+		return errors.Wrapf(err, "failed to create the %s %s for MachineDeployment %s", md.infrastructureMachineTemplate.GetKind(), md.infrastructureMachineTemplate.GetName(), md.object.GetName())
+	}
+
+	log.Info("creating", md.bootstrapTemplate.GroupVersionKind().String(), md.bootstrapTemplate.GetName())
+	if err := r.Client.Create(ctx, md.bootstrapTemplate); err != nil {
+		return errors.Wrapf(err, "failed to create the %s %s for MachineDeployment %s", md.bootstrapTemplate.GetKind(), md.bootstrapTemplate.GetName(), md.object.GetName())
+	}
+
+	log.Info("creating", md.object.GroupVersionKind().String(), md.object.GetName())
+	if err := r.Client.Create(ctx, md.object); err != nil {
+		return errors.Wrapf(err, "failed to create the %s %q", md.object.Kind, md.object.GetName())
+	}
+	return nil
+}
+
+func (r *ClusterTopologyReconciler) deleteMachineDeployment(ctx context.Context, md machineDeploymentTopologyState) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	log.Info("deleting", md.object.GroupVersionKind().String(), md.object.GetName())
+	if err := r.Client.Delete(ctx, md.object); err != nil && !apierrors.IsNotFound(err) {
+		return errors.Wrapf(err, "failed to delete the %s %q", md.object.Kind, md.object.Name)
+	}
+	return nil
+}
+
+func (r *ClusterTopologyReconciler) updateMachineDeployment(ctx context.Context, clusterName string, currentMD, desiredMD machineDeploymentTopologyState) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Check if the current and desired InfrastructureMachineTemplate are compatible.
+	if err := checkReferencedObjectsAreCompatible(currentMD.infrastructureMachineTemplate, desiredMD.infrastructureMachineTemplate); err != nil {
+		return err
+	}
+	// Check if the current and desired BoostrapTemplate are compatible.
+	// NOTE: this should never happen (webhooks prevent it), but checking for extra safety.
+	if currentMD.bootstrapTemplate.GetNamespace() != desiredMD.bootstrapTemplate.GetNamespace() {
```

Comment:
This check is already included in `checkReferencedObjectsAreCompatible`

Reply:
It is but I only need the namespace comparison and not everything else which is compared in `checkReferencedObjectsAreCompatible` in this case. (I also commented accordingly on your PR)